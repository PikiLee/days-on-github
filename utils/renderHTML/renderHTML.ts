import { readFile } from 'fs/promises'

import path from 'path'
import { render } from '../../dist/server/render.mjs'
import { AppProps } from '~/src/App'

export interface GithubData {
  daysOnGithub: number
  percentageDaysOnGithub: number
  contributionDays: {
    date?: string
    contributionCount?: number
  }[]
  name: string
  login: string
  avatarUrl: string | null
}

export async function renderHTML(props: AppProps) {
  const template = await readFile(
    path.join(process.cwd(), 'index.html'),
    'utf-8'
  )
  const css = await readFile(
    path.join(process.cwd(), './dist/output.css'),
    'utf-8'
  )
  const rendered = await render(props)

  const html = template
    .replace('<script type="module" src="./src/main.tsx"></script>', '')
    .replace(
      '<link rel="stylesheet" href="./src/index.css" />',
      `<style>${css}</style>`
    )
    .replace('<!--app-html-->', rendered ?? '')

  return html
}
