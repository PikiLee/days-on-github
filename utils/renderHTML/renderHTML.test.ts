import fsp from 'fs/promises'
import { describe, it, expect, vi } from 'vitest'
import { renderHTML, GithubData } from './renderHTML'

vi.mock('fs/promises')

vi.mocked(fsp.readFile)
  .mockResolvedValueOnce(
    `<!doctype html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Days on Github</title>
    <link rel="stylesheet" href="./src/index.css" />
  </head>
  <body>
    <div id="root"><!--app-html--></div>
    <script type="module" src="./src/main.tsx"></script>
  </body>
</html>
`
  )
  .mockResolvedValueOnce('body { background-color: white; }')

const mockRenderedContent = '<div>Github Stats</div>'
vi.mock('../../dist/server/render.mjs', () => ({
  render: vi.fn(() => mockRenderedContent)
}))

describe('renderHTML', () => {
  it('should render HTML with the provided Github data', async () => {
    const githubData = {} as GithubData

    const result = await renderHTML({ githubData })

    expect(result).toMatchSnapshot()
  })
})
