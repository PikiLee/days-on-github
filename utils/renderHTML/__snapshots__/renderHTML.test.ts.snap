// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`renderHTML > should render HTML with the provided Github data 1`] = `
"<!doctype html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Days on Github</title>
    <style>body { background-color: white; }</style>
  </head>
  <body>
    <div id="root"><div>Github Stats</div></div>
    
  </body>
</html>
"
`;
